File: extract-v20.js
const fs = require('fs');

const htmlContent = fs.readFileSync('soc_roster_final_v20.html', 'utf8');

const cssMatch = htmlContent.match(/<style>([\s\S]*?)<\/style>/);
const css = cssMatch ? cssMatch[1] : '';

const scriptMatch = htmlContent.match(/<script>([\s\S]*?)<\/script>/);
const js = scriptMatch ? scriptMatch[1] : '';

// The body is everything between </header> (or <body>) and <script>
// Looking at v20, let's just grab the whole body innerHTML
let bodyHtml = htmlContent.match(/<body[^>]*>([\s\S]*?)<script>/)[1];

// 1. Scoped CSS
let scopedCss = css.split('\n').map(line => {
  if (line.trim().startsWith('@') || line.trim() === '') return line;
  if (line.includes('{')) {
    // Basic prefixing (won't catch all edge cases but good enough for simple CSS)
    return '#rb-container ' + line.trim();
  }
  return line;
}).join('\n');
// We will just do a simpler CSS wrapper using standard CSS nesting or just keep it simple
// Wait, standard CSS doesn't support nesting without a preprocessor or modern CSS.
// Let's just write raw CSS, we'll manually namespace the few global tags like body, table.
let cleanCss = css.replace(/body\s*\{/g, '#rb-container {')
                  .replace(/table\s*\{/g, '#rb-container table {')
                  .replace(/th,\s*td\s*\{/g, '#rb-container th, #rb-container td {')
                  .replace(/button\s*\{/g, '#rb-container button {')
                  .replace(/select\s*\{/g, '#rb-container select {');

fs.writeFileSync('frontend/styles/roster-builder.css', cleanCss);

// 2. IIFE for JS
let cleanJs = `const RosterBuilder = (function() {
${js}

  // Expose initEmps and build
  return {
    initEmps: function(empMaster) {
       // mapping employeeMaster to EMPS_DEFAULT format
       emps = empMaster.map(e => ({
         id: e.slNo,
         name: e.name,
         dept: e.shiftGroup,
         type: e.role === 'SOC Analyst' ? 'Rotating' : 'General'
       }));
       renderEmpTable();
       resetCarryUI();
       initUI();
    },
    buildRoster: buildRoster
  };
})();`;

// Wait, the v20 script has "const EMPS_DEFAULT" and "let emps = ...". 
// We'll replace it.
cleanJs = cleanJs.replace(/const EMPS_DEFAULT = \[[\s\S]*?\];/, '');
cleanJs = cleanJs.replace(/let emps = JSON\.parse\(JSON\.stringify\(EMPS_DEFAULT\)\);/, 'let emps = [];');

fs.writeFileSync('frontend/modules/admin/roster-engine.js', cleanJs);

// 3. HTML 
fs.writeFileSync('frontend/modules/admin/roster-builder-ui.js', `
const RosterBuilderUI = \`
<div id="rb-container">
\${bodyHtml}
</div>
\`;
`);

console.log('Extraction complete');

File: extract-v20.py

import re

with open('soc_roster_final_v20.html', 'r', encoding='utf-8') as f:
    html = f.read()

css_match = re.search(r'<style>([\s\S]*?)</style>', html)
css = css_match.group(1) if css_match else ''

script_match = re.search(r'<script>([\s\S]*?)</script>', html)
js = script_match.group(1) if script_match else ''

body_match = re.search(r'<body[^>]*>([\s\S]*?)<script>', html)
body = body_match.group(1) if body_match else ''

# Clean CSS
lines = css.split('\n')
out_lines = []
for line in lines:
    if line.strip().startswith('@') or not line.strip():
        out_lines.append(line)
        continue
    if '{' in line:
        out_lines.append('#rb-container ' + line.strip())
    else:
        out_lines.append(line)

clean_css = '\n'.join(out_lines)
clean_css = clean_css.replace('body {', '#rb-container {')
clean_css = clean_css.replace('table {', '#rb-container table {')

with open('frontend/styles/roster-builder.css', 'w', encoding='utf-8') as f:
    f.write(clean_css)

# Clean JS
# Remove EMPS_DEFAULT and let emps =
js = re.sub(r'const EMPS_DEFAULT\s*=\s*\[[\s\S]*?\];', '', js)
js = re.sub(r'let emps\s*=\s*JSON\.parse\(JSON\.stringify\(EMPS_DEFAULT\)\);', 'let emps = [];', js)

clean_js = f"""const RosterBuilder = (function() {{
{js}

  return {{
    initEmps: function(empMaster) {{
       emps = empMaster.map(e => ({{
         id: e.slNo,
         name: e.name,
         dept: e.shiftGroup,
         type: e.role === 'SOC Analyst' ? 'Rotating' : 'General'
       }}));
       renderEmpTable();
       resetCarryUI();
       initUI();
    }},
    buildRoster: buildRoster
  }};
}})();"""

with open('frontend/modules/admin/roster-engine.js', 'w', encoding='utf-8') as f:
    f.write(clean_js)

# HTML UI
ui_js = f"""const RosterBuilderUI = `
<div id="rb-container">
{body}
</div>
`;
"""
with open('frontend/modules/admin/roster-builder-ui.js', 'w', encoding='utf-8') as f:
    f.write(ui_js)

print("Extraction complete.")
